core/include/mutex.h

```diff
@@ -8,116 +8,10 @@
- *
- * Obtaining a Mutex
- * -----------------
- *
- * If a `mutex_lock()` is called, one of the following happens:
- *
- * 1. If the mutex was unlocked (value of `NULL`), its value is changed to
- *    `MUTEX_LOCKED` and the call to `mutex_lock()` returns right away without
- *    blocking.
- * 2. If the mutex has a value of `MUTEX_LOCKED`, it will be changed to point to
- *    the `thread_t` of the running thread. The single item list is terminated
- *    by setting the `thread_t::rq_entry.next` of the running thread to `NULL`.
- *    The running thread blocks as described below.
- * 3. Otherwise, the current thread is inserted into the list of waiting
- *    threads sorted by thread priority. The running thread blocks as described
- *    below.
- *
- * In case 2) and 3), the running thread will mark itself as blocked (waiting
- * for a mutex) and yields. Once control is transferred back to this thread
- * (which is done in the call to `mutex_unlock()`), it has the mutex and the
- * function `mutex_lock()` returns.
- *
- * Returning a Mutex
- * -----------------
- *
- * If `mutex_unlock()` is called, one of the following happens:
- *
- * 1. If the mutex was already unlocked (value of `NULL`), the call returns
- *    without modifying the mutex.
- * 2. If the mutex was locked without waiters (value of `MUTEX_LOCKED`), it is
- *    unlocked by setting its value to `NULL`.
- * 3. Otherwise the first `thread_t` from the linked list of waiters is removed
- *    from the list.
- *     - This thread is the one with the highest priority, as the list is sorted
- *       by priority.
- *     - This thread's status is set to pending and its added to the appropriate
- *       run queue.
- *     - If that thread was the last item in the list, the mutex is set to
- *       `MUTEX_LOCK`.
- *     - The scheduler is run, so that if the unblocked waiting thread can
- *       run now, in case it has a higher priority than the running thread.
- *
- * Debugging deadlocks
- * -------------------
- *
- * The module `core_mutex_debug` can be used to print on whom `mutex_lock()`
- * is waiting. This information includes the thread ID of the owner and the
- * program counter (PC) from where `mutex_lock()` was called. Note that the
- * information is only valid if:
- *
- * - The mutex was locked by a thread, and not e.g. by `MUTEX_INIT_LOCKED`
- * - The function `cpu_get_caller_pc()` is implemented for the target
- *   architecture. (The thread ID will remain valid, though.)
- * - The caller PC is briefly 0 when the current owner passes over ownership
- *   to the next thread, but that thread didn't get CPU time yet to write its
- *   PC into the data structure. Even worse, on architectures where an aligned
- *   function-pointer-sized write is not atomic, the value may briefly be
- *   bogus. Chances are close to zero this ever hits and since this only
- *   effects debug output, the ostrich algorithm was chosen here.
- *
  * @{
- *
+ * @ingroup     core_sync_mutex
```

Comment:
```suggestion
 * @addtogroup core_sync_mutex
 * @{
```

The reason for the behavior observed by @mguetschow lies in the priority of the grouping commands: https://www.doxygen.nl/manual/grouping.html

```
The priorities of grouping definitions are (from highest to lowest): \ingroup, \defgroup, \addtogroup, \weakgroup.
```

The `@ingroup` command has a higher priority than `@defgroup`, which means that the contents of the header are not added to the group defined by `@defgroup` in the `mutex.doc.md` file. However the contents of the Markdown file can't be added to the header either, therefore Doxygen only creates the separate documentation page for the header.

The opening brackets should be below the command as well, but that is not the cause of the issue.

Reply:
Great, that did the trick: https://ci.riot-os.org/results/26628cecaa754671bd7631723421d236/doc-preview/group__core__sync__mutex.html

Reply:
defgroup ^= create a new group and put stuff inside 
ingroup ^= put stuff in group 

defgroup, ingroup ^= create a new group put it inside another group and put stuff into the new (sub)group

addtogroup ^= defgroup but do not care if already existing (basically fallback to ingroup) (defgroup with same label has priority for title and placement of that group)
weakgroup ^= addtogroup but this title has even less priority for the label and placement

The priotity is just within one doc-comment

search for ingroup -> place there; search for defgroup -> create new group and place there.

Reply:
> It should not show a different behavior

It's not entirely unlikely that there's a bug in Doxygen in that regard...

> The priotity is just within one doc-comment

..but I didn't find any reference in the Doxygen grouping documentation that this is the case. Or I missed it.

Reply:
```
To define a group, you should put the [\defgroup](https://www.doxygen.nl/manual/commands.html#cmddefgroup) command in a special comment block. The first argument of the command is a label that should uniquely identify the group. The second argument is the name or title of the group as it should appear in the documentation.
```
me :`defgroup ^= create a new group and put stuff inside`


```
You can make an entity a member of a specific group by putting a [\ingroup](https://www.doxygen.nl/manual/commands.html#cmdingroup) command inside its documentation block.
```
me: `ingroup ^= put stuff in group`

```
To avoid putting [\ingroup](https://www.doxygen.nl/manual/commands.html#cmdingroup) commands in the documentation for each member you can also group members together by the open marker @{ before the group and the closing marker @} after the group. The markers can be put in the documentation of the group definition or in a separate documentation block.
```
me: doxygen should have applied the ingroup label to everything in that `@{` group but it somehow created a subgroup 

```
If you don't want Doxygen to enforce unique labels, then you can use [\addtogroup](https://www.doxygen.nl/manual/commands.html#cmdaddtogroup) instead of [\defgroup](https://www.doxygen.nl/manual/commands.html#cmddefgroup). It can be used exactly like [\defgroup](https://www.doxygen.nl/manual/commands.html#cmddefgroup), but when the group has been defined already, then it silently merges the existing documentation with the new one.
```
me: `addtogroup is defgroup with ingroup fallback`

since non of this steps does nothing about priorities the priorities thing must be something else which is the evaluation order with in one step of placing a thing in the the documentation it goes ingroup then defgroup then add/weakgroup 

so it does not care whether one writes 

```
@ingroup in
@defgroup def
@addgroup add
```
or
 ```
@defgroup def
@addgroup add
@ingroup in
```

both will result in `in>def>add` in the group tree

Reply:
> > To avoid putting [\ingroup](https://www.doxygen.nl/manual/commands.html#cmdingroup) commands in the documentation for each member you can also group members together by the open marker @{ before the group and the closing marker @} after the group. The markers can be put in the documentation of the group definition or in a separate documentation block.
> 
> doxygen should have applied the ingroup label to everything in that `@{` group but it somehow created a subgroup

I think `@{ ... @}` just starts a new group (without giving it a name) which is then put inside another group with `@ingroup`. So that's consistent with what we saw.

Reply:
you can see this not happening in the sys_event documentation (all the different headers are added to the same group using an ingroup statement) and there is also no subgroup with addtogroup

and the thing i cited is from the doxygen documentation 

> To avoid putting [\ingroup](https://www.doxygen.nl/manual/commands.html#cmdingroup) commands in the documentation -- as in thevfollowing will have the same effect-- for each member you can also group members together by the open marker @{

Reply:
> you can see this not happening in the sys_event documentation (all the different headers are added to the same group using an ingroup statement) and there is also no subgroup with addtogroup

Ah now I see what you mean, different semantics if `@defgroup` is followed by `@{` vs. some `@ingroup` is followed by `@{` (as in if you split the `@defgroup` out to a `doc.md`). But I guess in that case you actually need `@addtogroup` then. As in `@defgroup` implies `@addtogroup`, but `@ingroup` doesn't.

All the other header files listed on the [Event Queue API page](https://doc.riot-os.org/group__sys__event.html#files) have none of their symbols listed on that same page directly, but only on the header file pages, e.g., https://doc.riot-os.org/callback_8h.html. Those implicit header file groups have no name (I would expect this example to be `sys_event_callback` for example).